pfunk/_resultsdb.py

```diff
@@ -0,0 +1,162 @@
+    For compatibility with the interfaces supplied by ResultsWriter/ResultsReader,
+    the flat-file equivalents, instances of this class accept a test name and date,
+    and provide access to the values in the row matching those properties only.
+    """
+
+    def __init__(self, filename, test_name, date):
+        self._connection = connect_to_database(filename)
+        self.__ensure_schema()
+        self._filename = filename
+        self._name = test_name
+        self._date = date
+        self.__ensure_row()
+
+    def __ensure_schema(self):
+        query = """ create table if not exists test_results(
+        identifier integer primary key asc,
+        name varchar,
+        date date,
+        status varchar,
+        python varchar,
+        pints varchar,
+        pints_commit varchar,
+        pfunk_commit varchar,
+        commit_hashes varchar,
+        seed integer,
+        method varchar,
+        json varchar
+        )"""
+        self._connection.execute(query)
+        self._connection.commit()
+
+    def __ensure_row(self):
+        # ensure the row exists
+        self._connection.execute("insert into test_results(name,date) values (?,?)", (self._name, self._date))
+        self._connection.commit()
+        rowID = self._connection.execute("select identifier from test_results where name like ? and date = ?",
+                                        (self._name, self._date))
+        self._row = rowID.fetchone()[0]
+
+    def __setitem__(self, key, value):
+        if key in self.primary_columns:
+            # don't update these
+            pass
+        elif key in self.mapped_columns.keys():
+            self[self.mapped_columns[key]] = value
+        elif key in self.columns:
+            self._connection.execute("update test_results set {} = ? where identifier = ?".format(key),
+                                    (value, self._row))
+            self._connection.commit()
+        else:
+            dictionary = self.json_values()
+            # workaround: if we're given a numpy array, make a Python list
+            if getattr(value, "tolist", None) is not None:
+                value = value.tolist()
+            dictionary[key] = value
+            json_field = json.dumps(dictionary)
+            self._connection.execute("update test_results set json = ? where identifier = ?",
+                                     (json_field, self._row))
+            self._connection.commit()
+
+    def write(self):
+        pass
+
+    def filename(self):
+        return self._filename
+
+
+class ResultsDatabaseReader(ResultsDatabaseSchemaClient):
+    def __init__(self, connection, rowID):
+        self._connection = connection
+        self._row = rowID
+
+    def __getitem__(self, item):
+        if item in self.primary_columns or item in self.columns:
+            result = self._connection.execute("select {} from test_results where identifier = ?".format(item),
```

Comment:
Do we need a `LIMIT 1` or something here? or is the identifier a key?

Reply:
Yes, it is:

    identifier integer primary key asc,